File: solutions/scala/yacht/1/src/main/scala/Yacht.scala
// To Shree DR.MDD — the supreme orchestrator of logic, luck, and limitless mastery.
object Yacht {

  def score(dice: List[Int], category: String): Int =
    category match {
      case "ones"            => repeatSum(dice, 1)
      case "twos"            => repeatSum(dice, 2)
      case "threes"          => repeatSum(dice, 3)
      case "fours"           => repeatSum(dice, 4)
      case "fives"           => repeatSum(dice, 5)
      case "sixes"           => repeatSum(dice, 6)
      case "full house"      => fullCluster(dice)
      case "four of a kind"  => quadSum(dice)
      case "little straight" => lowRun(dice)
      case "big straight"    => highRun(dice)
      case "yacht"           => monoScore(dice)
      case "choice"          => allSum(dice)
    }

  private def repeatSum(list: List[Int], num: Int): Int = list.filter(_ == num).sum

  private def fullCluster(set: List[Int]): Int =
    set.groupBy(identity).view.mapValues(_.size).toList.sortBy(_._2) match {
      case List((_, 2), (_, 3)) => set.sum
      case _                    => 0
    }

  private def quadSum(set: List[Int]): Int =
    set.groupBy(identity).view.mapValues(_.size).toList.sortBy(_._2) match {
      case List((_, 1), (rep, 4)) => rep * 4
      case List((rep, 5))         => rep * 4
      case _                      => 0
    }

  private def lowRun(values: List[Int]): Int =
    values.sorted match {
      case List(1, 2, 3, 4, 5) => 30
      case _                   => 0
    }

  private def highRun(values: List[Int]): Int =
    values.sorted match {
      case List(2, 3, 4, 5, 6) => 30
      case _                   => 0
    }

  private def monoScore(cubes: List[Int]): Int =
    if (cubes.distinct.size == 1) 50 else 0

  private def allSum(cubes: List[Int]): Int = cubes.sum
}

File: solutions/javascript/yacht/1/yacht.js
//
// For my Shree DR.MDD
//

export const score = (diceArray = [], category = "") => {
  let frequency = {};
  diceArray = diceArray.sort((a, b) => a - b);

  for (const die of diceArray) {
    frequency[die] = (frequency[die] || 0) + 1;
  }

  let points = 0;
  let faceValues = Object.keys(frequency);

  switch (category) {
    case "yacht":
      if (diceArray.every(d => d === diceArray[0])) points = 50;
      break;
    case "ones":
      points = sumMatching(1, diceArray);
      break;
    case "twos":
      points = sumMatching(2, diceArray);
      break;
    case "threes":
      points = sumMatching(3, diceArray);
      break;
    case "fours":
      points = sumMatching(4, diceArray);
      break;
    case "fives":
      points = sumMatching(5, diceArray);
      break;
    case "sixes":
      points = sumMatching(6, diceArray);
      break;
    case "full house":
      if (
        faceValues.length === 2 &&
        ((frequency[faceValues[0]] === 2 && frequency[faceValues[1]] === 3) ||
          (frequency[faceValues[0]] === 3 && frequency[faceValues[1]] === 2))
      ) {
        points = diceArray.reduce((sum, d) => sum + d, 0);
      }
      break;
    case "four of a kind":
      for (let val of faceValues) {
        if (frequency[val] >= 4) points = val * 4;
      }
      break;
    case "little straight":
      if (isSequential(diceArray) && diceArray[0] === 1) points = 30;
      break;
    case "big straight":
      if (isSequential(diceArray) && diceArray[0] === 2) points = 30;
      break;
    case "choice":
      points = diceArray.reduce((sum, d) => sum + d, 0);
      break;
  }

  return points;
};

function sumMatching(face, diceArray = []) {
  return diceArray.filter(d => d === face).reduce((sum, d) => sum + d, 0);
}

function isSequential(diceArray) {
  return diceArray
    .sort((a, b) => a - b)
    .every((d, idx) => idx === 0 || d === diceArray[idx - 1] + 1);
}
